File: packages/ogdf.js/src/layouts/energybased/sekk.js
import { PARAMETER_TYPE } from '../../utils/parameters'
import createLayout from '../creater'

const NAME = 'SEKK'
const OUR_PARAMETERS = {
    useWorker: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: false,
    },
}

const ORIGIN_PARAMETERS = {
    computeMaxIterations: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: true,
    },
    desLength: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [-Infinity, Infinity],
        default: 0.0,
    },
    globalIterationFactor: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 16,
    },
    maxGlobalIterations: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 2147483647,
    },
    maxLocalIterations: {
        type: PARAMETER_TYPE.INT,
        range: [0, Infinity],
        default: 2147483647,
    },
    stopTolerance: {
        type: PARAMETER_TYPE.DOUBLE,
        range: [0, Infinity],
        default: 0.001,
    },
    useLayout: {
        type: PARAMETER_TYPE.BOOL,
        range: [true, false],
        default: true,
    },
    zeroLength: {
        type: PARAMETER_TYPE.BOOL,
        range: [-Infinity, Infinity],
        default: -1.0,
    },
}

const ATTRIBUTES = {
    node: [
        {
            name: "nodesX",
            mapper: node => node.x,
            type: PARAMETER_TYPE.DOUBLE
        },
        {
            name: "nodesY",
            mapper: node => node.y,
            type: PARAMETER_TYPE.DOUBLE
        },
        {
            name: "nodesWidth",
            mapper: node => "width" in node ? node.width : 1.,
            type: PARAMETER_TYPE.DOUBLE
        },
        {
            name: "nodesHeight",
            mapper: node => "height" in node ? node.height : 1.,
            type: PARAMETER_TYPE.DOUBLE
        }
    ],
    link: [],
    sequence: ['nodesX', "nodesY", "nodesWidth", "nodesHeight"]
}

const sekk = createLayout(NAME, OUR_PARAMETERS, ORIGIN_PARAMETERS, ATTRIBUTES)
export default sekk

File: packages/ogdf.js/src/entry/energybased/sekk.cpp
#include <ogdf/energybased/SpringEmbedderKK.h>
#include "../main.h"

EM_PORT_API(float*) SEKK(int node_num, int link_num, int* source, int* target, double* nodesX, double* nodesY, double* nodesWidth, double* nodesHeight, bool computeMaxIterations, double desLength, int globalIterationFactor, int maxGlobalIterations, int maxLocalIterations, double stopTolerance, bool useLayout, double zeroLength) {
	node* nodes;
	Graph G;
	GraphAttributes GA(G, GraphAttributes::nodeGraphics | GraphAttributes::edgeGraphics);

	nodes = new node[node_num];
	for (int i = 0; i < node_num; i++){
		nodes[i] = G.newNode();
		GA.x(nodes[i]) = nodesX[i];
		GA.y(nodes[i]) = nodesY[i];
		GA.width(nodes[i]) = nodesWidth[i];
		GA.height(nodes[i]) = nodesHeight[i];
	}

	edge e;

	for (int i = 0; i < link_num; i++) {
		e = G.newEdge(nodes[source[i]], nodes[target[i]]);
		GA.bends(e);
	}
	//LayoutModule
	SpringEmbedderKK *model = new SpringEmbedderKK();
	
	model->computeMaxIterations(computeMaxIterations);
    model->setDesLength(desLength);
    model->setGlobalIterationFactor(globalIterationFactor);
    model->setMaxGlobalIterations(maxGlobalIterations);
    model->setMaxLocalIterations(maxLocalIterations);
    model->setStopTolerance(stopTolerance);
    model->setUseLayout(useLayout);
    model->setZeroLength(zeroLength);

    model->call(GA);

	float* re = (float*)malloc(node_num * 2 * 4);
    for(int i = 0; i < node_num; ++i) {
        re[i * 2] = GA.x(nodes[i]);
        re[i * 2 + 1] = GA.y(nodes[i]);
    }

    return re;
}